### PR TITLE
Refactor create wallet

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/hooks/index.ts [hunk 1]
```diff
@@ -565,7 +565,7 @@ export const useOpenWallet = (options?: UseMutationOptions<[YoroiWallet, WalletM
   const walletManager = useWalletManager()
   const mutation = useMutation({
     ...options,
-    mutationFn: async (walletMeta) => walletManager.openWallet(walletMeta),
+    mutationFn: (walletMeta) => Promise.all([walletManager.openWallet(walletMeta), walletMeta]),
```
Comment (michalszorad) [inline src/hooks/index.ts:568]: Is the `Promise.all` needed here?

**Review by michalszorad — APPROVED** [review]
I added one minor comment, but otherwise, the changes look good to me

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 2]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+      internalChain,
+      externalChain,
+    })
 
+    await wallet.discoverAddresses()
+    wallet.setupSubscriptions()
+    wallet.notify({type: 'initialize'})
+    wallet.isInitialized = true
+
+    if (!isYoroiWallet(wallet)) throw new Error('invalid wallet')
```
Comment (banklesss) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:301]: Should it be added before notifying the `initialize` event and setting `isInitialized` as `true`?

File: src/hooks/index.ts (hunk 1 above)
Comment (stackchain) [reply src/hooks/index.ts:568]: It is just to return the `walletMeta` alltogheter

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 3]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
-    this.utxoManager = utxoManager
-    this._utxos = utxoManager.initialUtxos
-    this.encryptedStorage = makeWalletEncryptedStorage(id)
-  }
+  static async create({
+    id,
+    networkId,
+    implementationId,
+    storage,
+
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:143]:
```suggestion
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 4]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+    mnemonic,
+    password,
+  }: {
+    id: string
+    implementationId: WalletImplementationId
+    networkId: NetworkId
+    storage: typeof storageLegacy
+    provider: YoroiProvider | undefined
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:151]: maybe is missing, but we can remove it
```suggestion
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 5]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:152]:
```suggestion
```

File: src/legacy/storage.ts [hunk 6]
```diff
@@ -79,6 +79,15 @@ export const keys = async (path: string, includeSubdirs?: boolean): Promise<Arra
   }
 }
 
+const storage = {
```
Comment (stackchain) [inline src/legacy/storage.ts:82]: maybe export the typeof

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 7]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+      provider: undefined,
+      accountPubKeyHex,
+      rewardAddressHex,
+      transactionCache,
+      internalChain,
+      externalChain,
+    })
 
+    await wallet.discoverAddresses()
+    wallet.setupSubscriptions()
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:296]: `discoveryAddresses` fire some events dunno about this order. I guess is ok

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 8]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
-    // Create at least one address in each block
-    await this.internalChain.initialize()
-    await this.externalChain.initialize()
-
-    await this.discoverAddresses()
-
     this.setupSubscriptions()
     this.notify({type: 'initialize'})
-
     this.isInitialized = true
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:361]: ? maybe
```suggestion
    this.initialized = true
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 9]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+    externalChain,
+  }: {
+    storage: typeof storageLegacy
+    networkId: NetworkId
+    id: string
+    utxoManager: UtxoManager
+    implementationId: WalletImplementationId
+    hwDeviceInfo: HWDeviceInfo | null
+    isReadOnly: boolean
+    provider: YoroiProvider | null | undefined
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:329]:
```suggestion
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 10]
```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
...
+
+  private constructor({
+    storage,
+    networkId,
+    id,
+    utxoManager,
+    implementationId,
+    hwDeviceInfo,
+    isReadOnly,
+    provider,
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:313]:
```suggestion
```

File: src/yoroi-wallets/cardano/types.ts [hunk 11]
```diff
@@ -72,15 +70,8 @@ export interface WalletInterface {
   // and was never updated. Starting from v4.1.0, we instead store the
   // last version the wallet has been *opened* on, since this is the actual
   // relevant information we need to decide on whether migrations are needed.
-  version: undefined | string
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:75]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 12]
```diff
@@ -72,15 +70,8 @@ export interface WalletInterface {
...
-
-  state: WalletState
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:77]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 13]
```diff
@@ -72,15 +70,8 @@ export interface WalletInterface {
...
-
-  isInitialized: boolean
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:79]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 14]
```diff
@@ -72,15 +70,8 @@ export interface WalletInterface {
...
-
-  transactionCache: null | TransactionCache
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:81]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 15]
```diff
@@ -102,30 +93,10 @@ export interface WalletInterface {
 
   get receiveAddresses(): Addresses
 
-  // =================== create =================== //
-
-  create(
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:107]: moved to static

File: src/yoroi-wallets/cardano/types.ts [hunk 16]
```diff
@@ -102,30 +93,10 @@ export interface WalletInterface {
...
-    mnemonic: string,
-    newPassword: string,
-    networkId: NetworkId,
-    implementationId: WalletImplementationId,
-    provider?: null | YoroiProvider,
-  ): Promise<string>
-
-  createWithBip44Account(
-    accountPublicKey: string,
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:116]: moved to static

File: src/yoroi-wallets/cardano/types.ts [hunk 17]
```diff
@@ -151,19 +120,10 @@ export interface WalletInterface {
   // TODO: type
   toJSON(): unknown
 
-  restore(data: unknown, walletMeta: WalletMeta): Promise<void>
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:154]: moved to static

File: src/yoroi-wallets/cardano/types.ts [hunk 18]
```diff
@@ -138,8 +109,6 @@ export interface WalletInterface {
 
   // =================== synch =================== //
 
-  doFullSync(): Promise<void>
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:141]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 19]
```diff
@@ -151,19 +120,10 @@ export interface WalletInterface {
   // TODO: type
   toJSON(): unknown
 
-  restore(data: unknown, walletMeta: WalletMeta): Promise<void>
-
   // =================== tx building =================== //
 
-  // not exposed to wallet manager, consider removing
-  getChangeAddress(): string
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:159]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 20]
```diff
@@ -151,19 +120,10 @@ export interface WalletInterface {
...
-
   getAllUtxosForKey(): Promise<Array<CardanoTypes.CardanoAddressedUtxo>>
 
-  getAddressing(address: string): unknown
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:163]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 21]
```diff
@@ -151,19 +120,10 @@ export interface WalletInterface {
...
-
-  getAddressedUtxos(): Promise<Array<CardanoTypes.CardanoAddressedUtxo>>
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:165]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 22]
```diff
@@ -322,7 +281,8 @@ type YoroiWalletKeys =
   | 'canGenerateNewReceiveAddress'
   | 'generateNewReceiveAddressIfNeeded'
   | 'generateNewReceiveAddress'
-  | 'receiveAddresses'
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:325]: duplicate

File: src/yoroi-wallets/cardano/types.ts [hunk 23]
```diff
@@ -288,7 +248,6 @@ type YoroiWalletKeys =
   | 'createVotingRegTx'
   | 'createWithdrawalTx'
   | 'disableEasyConfirmation'
-  | 'doFullSync'
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:291]: private

File: storybook/mocks/wallet.ts [hunk 24]
```diff
@@ -171,8 +171,11 @@ const wallet: YoroiWallet = {
   save: async (...args) => {
     action('save')(...args)
   },
-  doFullSync: async (...args) => {
```
Comment (wolverineks) [inline storybook/mocks/wallet.ts:174]: private

File: src/yoroi-wallets/cardano/types.ts [hunk 25]
```diff
@@ -102,30 +93,10 @@ export interface WalletInterface {
...
-    implementationId: WalletImplementationId,
-    hwDeviceInfo: null | HWDeviceInfo,
-    isReadOnly: boolean,
-  ): Promise<string>
-
   // ============ security & key management ============ //
 
   encryptAndSaveRootKey(encryptionMethod: EncryptionMethod, rootKey: string, password: string): Promise<void>
 
-  getDecryptedRootKey(rootPassword: string, intl: IntlShape): Promise<string>
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:127]: deprecated

Comment (wolverineks) [reply src/yoroi-wallets/cardano/types.ts:127]: used only in testing

File: src/yoroi-wallets/cardano/types.ts (hunk 11 above)
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:75]: deprecated

Comment (wolverineks) [reply src/yoroi-wallets/cardano/types.ts:75]: only used in testing

File: src/yoroi-wallets/cardano/types.ts [hunk 26]
```diff
@@ -43,27 +41,27 @@ export type WalletEvent =
 export type WalletSubscription = (event: WalletEvent) => void
 export type Unsubscribe = () => void
 export interface WalletInterface {
-  id: null | string
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:46]: no more `Nullable<T>`...

File: src/yoroi-wallets/cardano/ShelleyWallet.legacy.ts [hunk 27]
```diff
@@ -0,0 +1,1238 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.legacy.ts:1]: unchanged

File: src/hooks/index.ts (hunk 1 above)
Comment (wolverineks) [reply src/hooks/index.ts:568]: casual cleanup

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 4 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:151]: next pr

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 7 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:296]: its the same in the current ShelleyWallet version

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 9 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:329]: next pr

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 8 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:361]: lets keep it for now
most of the yoroi booleans are `isFoo` anyway

File: src/legacy/storage.ts (hunk 6 above)
Comment (wolverineks) [reply src/legacy/storage.ts:82]: for what?

**Review by stackchain — APPROVED** [review]
lgtm great job!

File: src/legacy/storage.ts (hunk 6 above)
Comment (stackchain) [reply src/legacy/storage.ts:82]: many places are extracting the type `typeof storage` DRY principle.

Comment (wolverineks) [reply src/legacy/storage.ts:82]: this is a pretty big change, we can do that later

Comment (stackchain) [reply src/legacy/storage.ts:82]: g2 agree

**Pull request merged**